### PR TITLE
prevent potential NPE on dispose

### PR DESCRIPTION
**Pull request by kaikreuzer**
Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

File: extensions/binding/org.eclipse.smarthome.binding.bluetooth/src/main/java/org/eclipse/smarthome/binding/bluetooth/ConnectedBluetoothHandler.java [hunk 1]
```diff
@@ -78,10 +78,14 @@ public void initialize() {
...
             connectionJob.cancel(true);
+            connectionJob = null;
         }
-        if (device != null) {
-            scheduler.submit(() -> device.disconnect());
-        }
+        scheduler.submit(() -> {
+            if (device != null) {
+                device.disconnect();
+                device = null;
```
Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.bluetooth/src/main/java/org/eclipse/smarthome/binding/bluetooth/ConnectedBluetoothHandler.java]: Do we care about threading?
`device` is set in the thread that calls `initialize` (in the parent class), `device` is unset (set to `null`) in a thread used by `scheduler`. So, should we use volatile, some synchronization, ...?

Comment (kaikreuzer) [reply extensions/binding/org.eclipse.smarthome.binding.bluetooth/src/main/java/org/eclipse/smarthome/binding/bluetooth/ConnectedBluetoothHandler.java]: You are right, I'll rework this...

Comment (kaikreuzer) [reply extensions/binding/org.eclipse.smarthome.binding.bluetooth/src/main/java/org/eclipse/smarthome/binding/bluetooth/ConnectedBluetoothHandler.java]: Using a reentrant lock now to make sure that different threads are correctly synchronized when accessing the device.

File: extensions/binding/org.eclipse.smarthome.binding.bluetooth/src/main/java/org/eclipse/smarthome/binding/bluetooth/BeaconBluetoothHandler.java [hunk 2]
```diff
@@ -67,18 +70,23 @@ public void initialize() {
         }
 
         adapter = (BluetoothAdapter) bridgeHandler;
+
+        deviceLock.lock();
```
Comment (sjka) [inline extensions/binding/org.eclipse.smarthome.binding.bluetooth/src/main/java/org/eclipse/smarthome/binding/bluetooth/BeaconBluetoothHandler.java]: locks should generally be guarded by `try {} finally {}`-blocks - especially when some (foreign) non-trivial code is called while the lock is in place.

Comment (kaikreuzer) [reply extensions/binding/org.eclipse.smarthome.binding.bluetooth/src/main/java/org/eclipse/smarthome/binding/bluetooth/BeaconBluetoothHandler.java]: thx. updated.

**Comment by maggu2810** [comment]
One question:
As we don't hold the lock over functions etc. wouldn't an sync. object (e.g. `deviceSync`) and a `synchronized (deviceSync) { ... }` simpler / safer (you cannot forget an unlock) then `try { ...lock(); ... } finally { ...unlock(); }`?

**Comment by kaikreuzer** [comment]
I had chosen the ReentrantLock in order to allow sub-classes to potentially do more complex things, if needed. I agree that with the current implementation as simply object would suffice for the synchronisation.

**Comment by maggu2810** [comment]
As #5303 has been merged now, should we (you) rebase your branch, so the build doesn't fail?

**Comment by kaikreuzer** [comment]
Which build fails...? All green for me :-)

**Comment by maggu2810** [comment]
I re-triggered the failed build immediately before I post the comment.

**Pull request merged**